game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneral.java

```diff
@@ -0,0 +1,135 @@
+package games.strategy.triplea.delegate.battle.steps.fire.general;
+
+import static games.strategy.triplea.delegate.battle.BattleState.Side.DEFENSE;
+import static games.strategy.triplea.delegate.battle.BattleState.Side.OFFENSE;
+import static games.strategy.triplea.delegate.battle.BattleState.UnitBattleFilter.ACTIVE;
+import static games.strategy.triplea.delegate.battle.BattleState.UnitBattleFilter.ALIVE;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.AIR_FIRE_NON_SUBS;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.steps.fire.FiringGroup;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import lombok.Value;
+import org.triplea.java.PredicateBuilder;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Create firing groups with units that match {@link #firingUnitPredicate}
+ *
+ * <p>The firing groups are separated by canNotTarget, canNotBeTargetedBy, and isSuicideOnHit
```

Comment:
It's a bit odd to have a generic predicate and for this class to know how the firing groups are going to be split.

Is the predicate acting as a filter to remove units from the firing groups, or is the predicate more acting at the firing group level?

The firing predicate is a bit concerning as I wonder if we are too abstract here, too complex. Could the caller apply the predicate rather than passing it in? I'm wondering if a caller would have to know a great deal about this implementation to be able to reasonably pass a good predicate to this class.

Reply:
Here is one thought, the list of split predicates is pretty well defined. What if we passed an enum value to specify the split type? This would I think be pretty beneficial as it would push complexity into this module, simplifying the callers. The cohesion of this class would be stronger, to understand what it's doing you would not have to look at the predicates being passed in, which is fundamentally a better and more useful module. Furthermore, by passing the right enum value, you get the right behavior, this really abstracts away implementation details from the callers. Otherwise, with a predicate a caller (or a maintainer, or someone looking to use or understand this code) has to have pretty deep knowledge of how the predicate is used.

Reply:
Do you mean the "filter type"?  Because the predicate is for filtering the units, it isn't for splitting the units.  But having an enum for the three different filters sounds like an ok idea.

Reply:
I've changed the predicate to enums that determine the predicate.